### ISSUE TITLE
Questions about SpanPredictor baseline

### ISSUE THREAD

**Issue by NoviScl** [shmsw25/AmbigQA#17]
Hi,

I'm trying to reproduce the SpanPredictor + Thresholding baseline but have some questions about the actual implementation:

1) In line 79 of span_predictor.py, you set `sel_labels = torch.zeros(N, dtype=torch.long).cuda()`. My understanding is that sel module predicts whether the passage contains a correct answer or not. If that's the case, why is sel_labels set to all zero?

2) I've browsed through some data and find that some passages contain multiple correct answer spans. In such cases, did you take into account all the detected spans during training? (i.e., each passage can have multiple different correct answer spans and we optimise likelihood over all of them instead of just one of them.)

3) How is the threshold selected exactly? I didn't seem to find the exact function / code snippet for selecting the threshold of predicting how many answers. (You used `np.log(0.05)` in AmbigQAData's evaluate function but I'm not sure where did that number come from.)

Thanks.

**Comment by shmsw25** [comment]
Hi @NoviScl, thanks for your question.

1. You are right that `sel_labels` indicate which passages contain a correct answer or not. They are always zeros because we made it that way - if you see [DataLoader](https://github.com/shmsw25/AmbigQA/blob/master/codes/DataLoader.py#L125), we sample one positive and M-1 negatives, and concatenate them. Therefore, the positive passage is always the zero-th passage. 
2. Yes, when we calculate the loss value [here](https://github.com/shmsw25/AmbigQA/blob/master/codes/models/span_predictor.py#L90), we use a maximum marginal log likelihood of the all possible spans.
3.  Good question. The value is chosen by tuning it as a hyper-parameter on the dev set. The code does not include the script to test it, but basically what we did is to use different values of thresholds and choose the one that yields the best result on the dev set.

**Comment by NoviScl** [comment]
Thanks a lot for the helpful reply!

**Comment by NoviScl** [comment]
Oh, just one follow-up question: For the SpanPredictor model, it's still not possible to output multiple answers for the same retrieved passage during inference right?

**Comment by shmsw25** [comment]
It actually can output multiple answers: if you see [util.py](https://github.com/shmsw25/AmbigQA/blob/master/codes/util.py#L92), there is an argument called `topk_answer`. For instance, if `topk_answer=3`, the inference algorithm detects at most top 3 spans per passage, and then further sort them based on likelihood. Therefore, the top answers may or may not be from the same passage.

**Comment by NoviScl** [comment]
Ah I see. It seems that `topk_answer=1` was set by default. But technically if we increase this number (i.e., obtaining more answer spans from each passage) the recall should be higher right? And if we further tune the `log_softmax` threshold properly, the overall F1 score could also be higher right? (i.e., both are tuneable hyper-parameters)

**Comment by shmsw25** [comment]
Sorry I missed your last question. Yes, I actually already tuned the threshold and chose the one that gave the best accuracy on the dev set.
